lib/datadog/tracing/distributed/datadog.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+# typed: true
+
+require_relative '../metadata/ext'
+require_relative '../trace_digest'
+require_relative 'datadog_tags_codec'
+
+module Datadog
+  module Tracing
+    module Distributed
+      # Datadog-style trace propagation.
+      class Datadog
+        def initialize(
+          trace_id: Ext::HTTP_HEADER_TRACE_ID,
+          parent_id: Ext::HTTP_HEADER_PARENT_ID,
+          sampling_priority: Ext::HTTP_HEADER_SAMPLING_PRIORITY,
+          origin: Ext::HTTP_HEADER_ORIGIN,
+          tags: Ext::HTTP_HEADER_TAGS,
+          fetcher: Fetcher
+        )
+          @trace_id = trace_id
+          @parent_id = parent_id
+          @sampling_priority = sampling_priority
+          @origin = origin
+          @tags = tags
+          @fetcher = fetcher
+        end
+
+        def inject!(digest, data)
+          return if digest.nil?
+
+          data[@trace_id] = digest.trace_id.to_s
+          data[@parent_id] = digest.span_id.to_s
+          data[@sampling_priority] = digest.trace_sampling_priority.to_s if digest.trace_sampling_priority
+          data[@origin] = digest.trace_origin.to_s unless digest.trace_origin.nil?
+
+          inject_tags(digest, data)
+
+          data
+        end
+
+        def extract(data)
+          fetcher = @fetcher.new(data)
+          trace_id = fetcher.id(@trace_id)
+          parent_id = fetcher.id(@parent_id)
+          sampling_priority = fetcher.number(@sampling_priority)
+          origin = fetcher[@origin]
+
+          # Return early if this propagation is not valid
+          # DEV: To be valid we need to have a trace id and a parent id
+          #      or when it is a synthetics trace, just the trace id.
+          # DEV: `Fetcher#id` will not return 0
+          return unless (trace_id && parent_id) || (origin && trace_id)
+
+          trace_distributed_tags = extract_tags(fetcher)
+
+          TraceDigest.new(
+            span_id: parent_id,
+            trace_id: trace_id,
+            trace_origin: origin,
+            trace_sampling_priority: sampling_priority,
+            trace_distributed_tags: trace_distributed_tags,
+          )
+        end
+
+        private
+
+        # Export trace distributed tags through the `x-datadog-tags` key.
+        #
+        # DEV: This method accesses global state (the active trace) to record its error state as a trace tag.
+        # DEV: This means errors cannot be reported if there's not active span.
+        # DEV: Ideally, we'd have a dedicated error reporting stream for all of ddtrace.
+        # DEV: The same comment applies to the {.extract_tags}.
```

Comment:
To be honest, this warning is important enough that I would suggest is actually worth copy-pasting into `#extract_tags`, rather than only referring to it here.

Reply:
Done.